app.go

```diff
@@ -757,20 +774,46 @@ func (app *App) start(ctx context.Context) error {
 
 			return multierr.Append(err, stopErr)
 		}
+
 		return err
 	}
 	app.log.LogEvent(&fxevent.Running{})
 
 	return nil
 }
 
-func withTimeout(ctx context.Context, f func(context.Context) error) error {
+type withTimeoutParams struct {
+	log       fxevent.Logger
+	hook      string
+	ctx       context.Context
+	callback  func(context.Context) error
+	lifecycle *lifecycleWrapper
+}
+
+func withTimeout(param *withTimeoutParams) error {
 	c := make(chan error, 1)
-	go func() { c <- f(ctx) }()
+	ctx := param.ctx
+	go func() { c <- param.callback(ctx) }()
 
 	select {
 	case <-ctx.Done():
-		return ctx.Err()
+		err := ctx.Err()
+		if err != context.DeadlineExceeded {
+			return err
+		}
+		// On timeout, report running hook's caller and recorded
+		// runtimes of hooks successfully run till end.
+		r := param.lifecycle.hookRecords()
+		caller := param.lifecycle.runningHookCaller()
+		if len(r) > 0 {
+			err = multierr.Append(err, fmt.Errorf(`
+timed out while executing hook %s (Caller: %s). Hooks successfully ran so far: %s`,
+				param.hook, caller, r))
+		} else {
+			err = multierr.Append(err, fmt.Errorf(`
+timed out while executing hook %s (Caller: %s)`, param.hook, caller))
+		}
```

Comment:
optionally, instead of appending, we can wrap:

```
type hookRunError struct {
  Records []HookRecord
  Err error
}

// (this is needed for errors.Is and errors.As APIs to function.)
// https://golang.org/pkg/errors/
func (e hookRunError) Unwrap() error  { return e.Err }

func (hookRunError) Error() string {
  // something like the string you're generating right now, maybe multi-line
}
```

This can handle both `len == 0` and `len > 0` cases in the Error string which can be independently tested.

---

Since ctx.Err will normally be one of cancelled or timeout, we can probably wrap the ctx.Err unconditionally.